### ISSUE TITLE
RFC: Changing the Octicons React API to support 24px icons

### ISSUE THREAD

**Issue by colebemis** [primer/octicons#393]
As part of the [Octicons refresh](https://github.com/github/design-systems/issues/711) project, we designed a set of 24px icons (in addition to 16px icons). With the addition of this 24px grid size, we no longer have a 1-to-1 mapping of icon name to SVG file. The Octicons React components now need to choose which SVG to render based on the size passed in:

| Before                                                                                                         | After                                                                                                         |
| -------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------- |
| ![image](https://user-images.githubusercontent.com/4608155/78410106-492bcf80-75c0-11ea-8af3-c7416b3ed2dc.png) | ![image](https://user-images.githubusercontent.com/4608155/78410127-56e15500-75c0-11ea-9ce7-0d9298b1c95b.png) |

The current public API of our Octicons React package makes implementing this change tricky. I'm proposing a simpler public API that would make this change easier to implement. Specifically, I'm purposing that we remove the `Octicon` "shell" component so each icon would be rendered with a single React component:

```diff
- <Octicon icon={Alert} />
+ <Alert />
```
I explain my reasoning for this API change below 👇 

## What is the public API right now?

We currently have 5+ ways to render an octicon in React:

1. `icon` prop (most common)

   ```jsx
   <Octicon icon={Alert} />
   ```

2. React `children`

   ```jsx
   <Octicon><Alert /></Octicon>
   ```

3. `getIconByName()` function

   ```jsx
   <Octicon icon={getIconByName('alert')} />
   ```

4. `iconsByName` object

   ```jsx
   <Octicon icon={iconByName.alert} />
   ```

5. `StyledOcticon` component

   ```jsx
   <StyledOcticon icon={Alert} />
   <StyledOcticon><Alert /></StyledOcticon>
   <StyledOcticon icon={getIconByName('alert')} />
   <StyledOcticon icon={iconsByName.alert} />
   ```

## Do we have to change the public API?

Probably. Let's see why.

We'll use this code example to anchor our discussion:

```jsx
<Octicon icon={Alert} size={24} />
```

The above snippet will render an `<svg>` on the page. Something like this:

```html
<svg ...>
  <path d="...">
</svg>
```

`Octicon` is responsible for rendering the `<svg>` "shell":

```html
<svg ...>

</svg>
```

`Alert` is responsible for rendering the elements inside the `<svg>`:

```html
<path d="...">
```

### Problem 1: Passing data from `Octicon` down to `Alert`

Since there is no longer a 1-to-1 mapping of icon name to SVG file, `Alert` now needs access to the `size` prop passed to `Octicon` in order to decide whether to render `alert-16.svg` or `alert-24.svg`.

We can solve this problem without changing the public API by using [context](https://reactjs.org/docs/context.html) to expose the `size` prop to `Alert`. That would look something like this:

```jsx
function Octicon({icon: Icon, size}) {
  return (
    <OcticonContext.Provider value={size}>
      <svg>
        <Icon />
      </svg>
    </OcticonContext.Provider>
  )
}

function Alert() {
   const heights = {
    16: {
      width: 16,
      path: '<path d="..." />'
    },
    24: {
      width: 24,
      path: '<path d="..." />'
    },
  }

  const size = React.useContext(OcticonSizeContext)
  const naturalHeight = chooseNaturalHeight(heights, size)
  const path = heights[naturalHeight].path

  return (
    <g dangerouslySetInnerHtml={{ __html: path }} />
  )
}
```

Okay, so we can use context to pass data from `Octicon` down to `Alert` without changing the public API. Awesome. However, we also need to be able to pass data from `Alert` back up to `Octicon`.

### Problem 2: Passing data from `Alert` up to `Octicon`

`Octicon` needs to know whether `Alert` chose to render `alert-16.svg` or `alert-24.svg` in order to compute the `viewBox` attribute.

Currently, the `Alert` component has a static `size` property that the `Octicon` component uses to compute the `viewBox`:

```jsx
function Alert() {
  return <path d="..." />
}

Alert.size = [16, 16]

function Octicon({icon: Icon, size}) {
  const [naturalWidth, naturalHeight] = Icon.type.size
  return (
    <svg viewBox={`0 0 ${naturalWidth} ${naturalHeight}`}>
      <Icon />
    </svg>
  )
}
```

However, this `size` property (which refers to the natural size of the SVG) is assigned to `Alert` at build-time and we no longer know what the natural size of `Alert` is at build-time. It could `[16, 16]` or `[24, 24]` depending on what the user passes in via the `size` prop. We now need a way to tell the `Octicon` component what the natural size of `Alert` is at _run-time_.

Let's extend our previous context implementation to pass that natural size information from `Alert` up to `Octicon`:

```diff
function Octicon({icon: Icon, size}) {
+ const [naturalWidth, setNaturalWidth] = React.useState(16)
+ const [naturalHeight, setNaturalHeight] = React.useState(16)
  return (
    <OcticonContext.Provider
      value={{
        size, 
+       setNaturalWidth, 
+       setNaturalHeight
      }}
    >
+     <svg viewBox={`0 0 ${naturalWidth} ${naturalHeight`}>
        <Icon />
      </svg>
    </OcticonContext.Provider>
  )
}

function Alert() {
   const heights = {
    16: {
      width: 16,
      path: '<path d="..." />'
    },
    24: {
      width: 24,
      path: '<path d="..." />'
    },
  }

  const {
    size,
+   setNaturalWidth,
+   setNaturalHeight,
  } = React.useContext(OcticonSizeContext)
  const naturalHeight = chooseNaturalHeight(heights, size)
+ const naturalWidth = heights[naturalHeight].width
  const path = heights[naturalHeight].path

+  React.useEffect(() => {
+   setNaturalWidth(naturalWidth)
+ }, [naturalWidth])

+ React.useEffect(() => {
+   setNaturalHeight(naturalHeight)
+ }, [naturalHeight])

  return (
    <g dangerouslySetInnerHtml={{ __html: path }} />
  )
}
```

Okay, this approach seems to work but it is not ideal. On the first render, we're guessing what the `viewBox` of the `<svg>` will be. If we're wrong, the icons will flash like this:

![Kapture 2020-04-03 at 15 30 59](https://user-images.githubusercontent.com/4608155/78410387-fe5e8780-75c0-11ea-8b53-58bb82467570.gif)

More generally, we seem to be creating an increasingly complex private API to preserve an already complex public API.

Since we'll be releasing new major versions of all the Octicons libraries as part of the refresh, why not use this opportunity to simplify the API?

## What should the new public API be?

I'm purposing that we remove the `Octicon` "shell" component. Each icon would be rendered with a single React component:

```diff
- <Octicon icon={Alert} />
+ <Alert />
```

Without the `Octicon` component, there will be no need to set up two-way data flow between `Octicon` and `Alert`, removing most of the complexity detailed above.

This also means that there will only be one way to render an octicon. No more `getIconByName()` or `iconsByName`.

This API is simliar to [`react-feather`](https://github.com/feathericons/react-feather)'s API which has been battle-tested by over 300 thousand npm installs per month.

### What about system props?

We could expose systems props through a separate companion package called, say, `@primer/octicons-react-styled`:

```jsx
import {Alert} from '@primer/octicons-react-styled'

<Alert color="gray.5" mr={2} />
```

### Implementation

Here's how this simpler API could be implemented:

```jsx
// @primer/octicons-react

function Alert({size = 16}) {
  const heights = {
    16: {
      width: 16,
      path: '<path d="..." />'
    },
    24: {
      width: 24,
      path: '<path d="..." />'
    },
  }

  const naturalHeight = chooseNaturalHeight(heights, size)
  const naturalWidth = heights[naturalHeight].width
  const path = heights[naturalHeight].path

  return (
    <svg
      viewBox={`0 0 ${naturalWidth} ${naturalHeight}`}
      dangerouslySetInnerHtml={{ __html: path }}
    />
  )
}
```

```jsx
// @primer/octicons-react-styled

const StyledAlert = styled(Alert)`
  ${COMMON}
`
```

## Open questions

- What are the consequences of removing `getIconByName()` and `iconsByName` from our API?
- Are there benefits to keeping the `Octicon` "shell" component that I'm not aware of?

---

I'd love to get feedback on this proposed API change from people using `@primer/octicons-react` in their projects. How would this change affect you?

**Comment by dmarcey** [comment]
Thanks for the detailed outline of the problem and options @colebemis !
Here are a few thoughts that jumped to mind.

1) It seems like it would be possible to keep the `<Octicon icon={Alert} size={16} />` API around if the shell were just something like this:
```
export const Octicon = ({Icon, size}) => {
  return <Icon size={size} />
}
```
Does that seem accurate? At this point we would just be leaving the `<Octicon />` component around for back-compat, because it would really would just be a wrapper around the individual icon components. I don't necessarily know that I would advocate for this as an option, I just thought I would mention it for completeness / check my understanding to make sure that it would be an option.

If we think that `<Octicon icon={Alert} />` is the most common usage of octicons in React today, then it would allow those consumers to not be broken. But, at the same time, it would leave around a superfluous component as well as a little bit of a fragmented API, so there are definitely pros/cons worth weighing.

2) One observation around the naming of the individual icon components. Should they have the word `Icon` somewhere in the name, i.e. `<AlertIcon />`? Maybe it is just the example that we're looking at here, but when I see `<Alert />` my mind immediately jumps to a an alert message or message bar type component. I know in `@primer/components` we call it a `<Flash />` (https://primer.style/components/Flash) but this is the type of component that I expect when I see something like `<Alert />` without some other indication that it is an icon. 
Looking through the library there are maybe only a handful of icons this problem would manifest, (I don't think I would have any trouble confusing `<LightBulb />` for a different type of component) so we should make sure we don't overcorrect on `Alert`, but it was something that did immediately jump out to me, so I thought it was worth mentioning.

Regarding the other open question, I've never personally used the `iconsByName` API, so I can't weigh in the negative implications of removing that option. However, I guess if we needed to keep the API (just as if we needed to keep the `<Octicon />` wrapper in general), we would be able to figure out a way to do so with the same sort of implementation that I included above.

Overall, I think the direction is good and from a consumer standpoint, I don't think reacting to the breaking change (should we decide to go that route) would be too disruptive in any code that I am a stakeholder for.

**Comment by dmarcey** [comment]
cc / @clayplumridge as another stakeholder for the Octions React package.

**Comment by colebemis** [comment]
Thanks for your comment, @dmarcey! 

1. I like your implementation idea for the `Octicon` component 😄 I hadn't thought of that before. Doing something like that would make for a much smoother transition. We could also add a deprecation warning in the `Octicon` component so we can remove it in the next major version:

```jsx
export const Octicon = ({Icon, ...props}) => {
  if (process.env.NODE_ENV === 'development) {
    console.warn("Octicon is deprecated ...")
  }
  return <Icon {...props} />
}
```

2. I agree that `<AlertIcon />` is more clear than `<Alert />` but I'm curious if the difference between the two is significant enough to warrant a breaking change. I'd love to hear other people's thoughts on this :)

**Comment by colebemis** [comment]
@dmarcey I got a prototype of that implementation working! https://codesandbox.io/s/young-waterfall-73imq

**Comment by dmarcey** [comment]
Makes sense to me. I extracted out a function that generates the `props` for the `svg`. That way the icon function can be as slim as possible - it now only has the `height` -> `{width, path}` object map in the icon function itself.

https://codesandbox.io/s/vibrant-kalam-ekhmk

**Comment by dmarcey** [comment]
Also, why do we need the `cloneElement` in `Octicon`? Is there a reason that this wouldn't work?
```
function Octicon({ icon: Icon, children, ...props }) {
  return typeof Icon === "function" ? <Icon {...props} /> : children;
}
```

**Comment by colebemis** [comment]
Even better. Thanks, @dmarcey!

**Comment by colebemis** [comment]
@dmarcey I feel like the `naturalHeights` object should have a different name but I'm not sure what to call it. Maybe `iconsByHeight` or something? What do you think?

**Comment by dmarcey** [comment]
`iconsByHeight` isn't bad. Maybe `pathDataByHeight`?

**Comment by colebemis** [comment]
Closed by #395
